lib/python/qmk/keymap.py

```diff
@@ -192,19 +199,79 @@ def generate_c(keyboard, layout, layers):
 
         layers
             An array of arrays describing the keymap. Each item in the inner array should be a string that is a valid QMK keycode.
+
+        macros
+            A sequence of strings containing macros to implement for this keyboard.
     """
-    new_keymap = template_c(keyboard)
+    new_keymap = template_c(keymap_json['keyboard'])
     layer_txt = []
-    for layer_num, layer in enumerate(layers):
+
+    for layer_num, layer in enumerate(keymap_json['layers']):
         if layer_num != 0:
             layer_txt[-1] = layer_txt[-1] + ','
         layer = map(_strip_any, layer)
         layer_keys = ', '.join(layer)
-        layer_txt.append('\t[%s] = %s(%s)' % (layer_num, layout, layer_keys))
+        layer_txt.append('\t[%s] = %s(%s)' % (layer_num, keymap_json['layout'], layer_keys))
 
     keymap = '\n'.join(layer_txt)
     new_keymap = new_keymap.replace('__KEYMAP_GOES_HERE__', keymap)
 
+    if keymap_json.get('macros'):
+        macro_txt = [
+            'bool process_record_user(uint16_t keycode, keyrecord_t *record) {',
+            '    if (record->event.pressed) {',
+            '        switch (keycode) {',
+        ]
+
+        for i, macro_array in enumerate(keymap_json['macros']):
+            macro = []
+            for macro_fragment in macro_array:
+                if isinstance(macro_fragment, str):
+                    macro.append('"' + macro_fragment.replace('"', r'\"') + '"')
```

Comment:
This needs to escape `\` too.

Supporting things like `\n` and `\t` could also be useful (`\t` may easily come with some copy/paste, and inserting multiline text using `\n` would be much easier than using `{"action":"tap","keycodes":["ENT"]}` for every line). Although raw tab characters would probably work anyway, raw newlines won't.

Reply:
I don't think we need to escape those here. At this point the backslashes are part of the text string, and will appear with a single backslash in the quoted C string. If we escape them here the macro will end up typing a literal `\n` instead of enter.

Reply:
So you want to pass backslashes to the C code, so that they work as escape characters in the resulting C string, but at the same time you escape double quotes? This creates some weird quoting rules.

I was thinking about not doing this kind of passthrough escaping at all — instead let JSON handle `\"`, `\\`, `\n`, `\t` (and maybe even `\b` for completeness), so that they will appear as raw characters in the string after JSON parsing, and then escape all those characters according to C rules when generating the code. This way you won't need to write `"\\n"` or `"\\\\"` in JSON to get just a single character in the resulting C string; also writing `"\\\""` or `"\n"` in JSON won't produce broken C code, as it would with your suggested implementation.

The only potential problem would be if you would want to have something like `\xA0` in the generated C string for some reason — but `send_string()` does not support anything above the 7-bit ASCII range anyway (`send_char()` does not even have a range check and would do undefined things if it gets a non-ASCII character), therefore such codes are not really needed there.

Reply:
I think you're misunderstanding the layer of abstraction here. Take the following JSON string:

    "Hello, \"$NAME\"\\n"

That will serialize down to the following string inside python:

    Hello, "$NAME"\n

And as a result when line 230 is executed the following will be written to the `keymap.c`:

    SEND_STRING("Hello, \"$NAME\"\n");

I am escaping double quotes because that is needed to get the correct string into C. We should not escape other things because they are already formatted correctly for their destination.

Reply:
So you propose that `"\\n"` in the JSON string results in `"\n"` in the C code? That's what I call “weird quoting rules” — why can't it be just `"\n"` in JSON, which gets translated to the actual newline character in Python, and then gets escaped to `"\n"` when writing C code? Similarly, `"\\"` in JSON would give a single backslash character in Python, which then would be escaped to `"\\"` when writing C code.

This won't allow `"\a"`, because it's not supported in JSON, but we have `{"action": "beep"}` for that.

Reply:
> why can't it be just `"\n"` in JSON, which gets translated to the actual newline character in Python, and then gets escaped to `"\n"` when writing C code?

I think I see what you're getting at. We can add code that translates literal new line characters (`\n`, 0x0A) and literal carriage returns (`\r`, 0x0D) to a literal `\n` inside the C string, it's probably just one or two lines of code. The only barrier was understanding that's what you wanted.

Reply:
32f62cf254 should do what you're asking for. I don't think we actually have to handle all 3 text formats, but given we're handling user input I'd rather be safe than sorry.

Reply:
Yes, and probably tab characters should be escaped to `\t` too (although technically they should work as raw characters in a C string, they might be confused with spaces if someone needs to look at the generated code). Most importantly, backslash characters should also be escaped, otherwise the confusion about the layer at which they are interpreted would remain.

In the end the string part of a macro would just be a plain string, and the fact that the underlying implementation generates C code to send that string would not be visible.

Reply:
> [32f62cf](https://github.com/qmk/qmk_firmware/commit/32f62cf254adc1a3733c82bfcc0fccfc66604199) should do what you're asking for.
No, this still seems to pass the backslash characters directly to the C string, so they can be interpreted as a start of a C-style escape sequence.

Reply:
> Yes, and probably tab characters should be escaped to `\t` too

Good idea.

> Most importantly, backslash characters should also be escaped, otherwise the confusion about the layer at which they are interpreted would remain.

I understand your point, but this also precludes users from getting those from JSON to C unless we've explicitly allowed for it. I need to think some more about this before I decide if I think that would be a feature or a bug.

Reply:
1949e456b7 adds quoting for backslashes.